api/clients/v2/verification/cert_verifier.go

```diff
@@ -219,105 +168,82 @@ func (cv *CertVerifier) GetNonSignerStakesAndSignature(
+	certVerifierCaller, err := cv.getVerifierCallerFromAddress(certVerifierAddress)
+	if err != nil {
+		return nil, fmt.Errorf("get verifier caller from address: %w", err)
 	}
 
 	quorumNumbersRequired, err := certVerifierCaller.QuorumNumbersRequiredV2(&bind.CallOpts{Context: ctx})
 	if err != nil {
 		return nil, fmt.Errorf("get quorum numbers required: %w", err)
 	}
 
+	cv.requiredQuorums.Store(certVerifierAddress, quorumNumbersRequired)
+
 	return quorumNumbersRequired, nil
 }
 
-// MaybeWaitForBlockNumber waits until the internal eth client has advanced to a certain targetBlockNumber, unless
-// configured pollInterval is <= 0, in which case this method will NOT wait for the internal client to advance.
+// getVerifierCallerFromBlockNumber returns a ContractEigenDACertVerifierCaller that corresponds to the input block
+// number
 //
-// This method will check the current block number of the internal client every CertVerifier.pollInterval duration.
-// It will return nil if the internal client advances to (or past) the targetBlockNumber. It will return an error
-// if the input context times out, or if any error occurs when checking the block number of the internal client.
+// If the eth node hasn't yet advanced to the input block number, this method will wait until that block is reached
+// before attempting to get the verifier caller.
 //
-// This method is synchronized in a way that, if called by multiple goroutines, only a single goroutine will actually
-// poll the internal eth client for most recent block number. The goroutine responsible for polling at a given time
-// updates an atomic integer, so that all goroutines may check the most recent block without duplicating work.
-func (cv *CertVerifier) MaybeWaitForBlockNumber(ctx context.Context, targetBlockNumber uint64) error {
-	if cv.pollInterval <= 0 {
-		// don't wait for the internal client to advance
-		return nil
-	}
-
-	if cv.latestBlockNumber.Load() >= targetBlockNumber {
-		// immediately return if the local client isn't behind the target block number
-		return nil
+// This method caches ContractEigenDACertVerifierCaller instances, since their construction requires acquiring a lock
+// and parsing json, and is therefore non-trivially expensive.
+func (cv *CertVerifier) getVerifierCallerFromBlockNumber(
+	ctx context.Context,
+	blockNumber uint64,
+) (*verifierBindings.ContractEigenDACertVerifierCaller, error) {
+	err := cv.blockNumberProvider.MaybeWaitForBlockNumber(ctx, blockNumber)
+	if err != nil {
+		return nil, fmt.Errorf("wait for block number: %w", err)
 	}
 
-	ticker := time.NewTicker(cv.pollInterval)
-	defer ticker.Stop()
-
-	polling := false
-	if cv.pollingActive.CompareAndSwap(false, true) {
-		// no other goroutine is currently polling, so assume responsibility
-		polling = true
-		defer cv.pollingActive.Store(false)
+	certVerifierAddress, err := cv.certVerifierAddressProvider.GetCertVerifierAddress(blockNumber)
```

Comment:
shouldnt the MaybeWaitForBlockNumber be done inside the certVerifierAddressProvider? Otherwise what if these 2 providers use a different underlying eth client that arent at the same block as an example

Reply:
Very nice. I definitely did not have such an complex change in mind but glad it helped you find this new abstraction :)